File: src/main/java/xyz/devvydont/smprpg/gui/items/MenuRecipeViewer.java
package xyz.devvydont.smprpg.gui.items;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Bukkit;
import org.bukkit.Keyed;
import org.bukkit.Material;
import org.bukkit.block.Campfire;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryOpenEvent;
import org.bukkit.inventory.*;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;
import org.jetbrains.annotations.NotNull;
import xyz.devvydont.smprpg.SMPRPG;
import xyz.devvydont.smprpg.gui.InterfaceUtil;
import xyz.devvydont.smprpg.gui.base.MenuBase;
import xyz.devvydont.smprpg.items.base.SMPItemBlueprint;
import xyz.devvydont.smprpg.items.interfaces.ICraftable;
import xyz.devvydont.smprpg.items.interfaces.ISmeltable;
import xyz.devvydont.smprpg.services.ItemService;
import xyz.devvydont.smprpg.services.RecipeService;
import xyz.devvydont.smprpg.util.formatting.ComponentUtils;
import xyz.devvydont.smprpg.util.time.TickTime;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class MenuRecipeViewer extends MenuBase {

    public static final int ROWS = 6;

    // Hard set positions
    public static final int CORNER = 10;
    public static final int RESULT = 23;
    public static final int REQUIREMENTS = 25;

    // The index of the recipe we want to show if there is more than one recipe.
    private int currentRecipe = 0;

    private final List<Recipe> recipes;
    private final ItemStack result;

    // Integer that will allow us to continuously flip through different options a recipe provides.
    private int recipeChoiceIndex = 0;

    /**
     * Default constructor initialized from within the MenuItemBrowser typically when a craftable item was clicked.
     *
     * @param player The player viewing the recipe.
     * @param parentMenu The menu calling this menu. Typically, the MenuItemBrowser.
     * @param result The result from crafting.
     */
    public MenuRecipeViewer(@NotNull Player player, MenuBase parentMenu, List<Recipe> recipes, ItemStack result) {
        super(player, ROWS, parentMenu);
        this.recipes = recipes;
        this.result = ItemService.clean(result);  // This may seem silly, but it's necessary to remove the "craftable" lore.
    }

    @Override
    protected void handleInventoryOpened(InventoryOpenEvent event) {
        super.handleInventoryOpened(event);
        event.titleOverride(ComponentUtils.merge(ComponentUtils.create("Recipes for: "), result.displayName()));
        this.render();
        Bukkit.getScheduler().runTaskTimer(SMPRPG.getInstance(), task -> {

            // If nobody is viewing us, we can stop the task.
            if (this.inventory.getViewers().isEmpty()) {
                task.cancel();
                return;
            }

            render();
            recipeChoiceIndex++;
        }, TickTime.HALF_SECOND, TickTime.HALF_SECOND);
    }

    @Override
    protected void handleInventoryClicked(InventoryClickEvent event) {
        super.handleInventoryClicked(event);
        event.setCancelled(true);
        this.playInvalidAnimation();
    }

    private ItemStack getItemFromRecipeChoice(RecipeChoice choice) {

        // Yes, this is possible... LMAO
        if (choice == null)
            return ItemStack.of(Material.AIR);

        ItemStack item = null;
        if (choice instanceof RecipeChoice.ExactChoice exact)
            item = ItemService.clean(exact.getChoices().get(recipeChoiceIndex % exact.getChoices().size()));

        if (choice instanceof RecipeChoice.MaterialChoice material)
            item = ItemService.generate(material.getChoices().get(recipeChoiceIndex % material.getChoices().size()));

        // This will only occur if Spigot/Paper ever add a new child of RecipeChoice in a future update.
        if (item == null) {
            SMPRPG.broadcastToOperatorsCausedBy(this.player, ComponentUtils.create("Unknown recipe choice candidate: " + choice.getClass().getSimpleName()));
            throw new IllegalStateException("Unknown choice type: " + choice.getClass().getSimpleName());
        }

        ItemService.blueprint(item).updateItemData(item);

        // If this item is deemed craftable by us, inject lore explaining that they can click it to go deeper.
        if (!RecipeService.getRecipesFor(item).isEmpty()) {
            var lore = new ArrayList<Component>(List.of(
                    ComponentUtils.EMPTY,
                    ComponentUtils.create("Click to view recipe!", NamedTextColor.YELLOW),
                    ComponentUtils.EMPTY
            ));
            if (item.lore() != null)
                lore.addAll(item.lore());
            item.lore(lore);
        }

        return item;
    }

    /**
     * When an ingredient is clicked, if the blueprint of that ingredient is craftable itself as well then open
     * another layer of this menu with its recipe.
     *
     * @param itemStack The ItemStack clicked as a crafting component
     */
    private void handleIngredientClick(ItemStack itemStack) {

        // Retrieve the blueprint of this item. If it is craftable, enter another recipe layer
        var recipesFor = RecipeService.getRecipesFor(ItemService.blueprint(itemStack).generate());
        if (recipesFor.isEmpty() || itemStack.getType().equals(Material.AIR)) {
            this.playInvalidAnimation();
            return;
        }

        this.openSubMenu(new MenuRecipeViewer(player, this, recipesFor, itemStack));
    }

    /**
     * Renders the recipe completely
     */
    public void renderRecipe() {

        if (recipes.isEmpty()) {
            SMPRPG.broadcastToOperatorsCausedBy(this.player, ComponentUtils.create("Achieved impossible recipe viewer state (no recipes to show): ", NamedTextColor.RED).append(result.displayName()));
            return;
        }

        if (currentRecipe >= recipes.size())
            currentRecipe = 0;
        if (currentRecipe < 0)
            currentRecipe = recipes.size() - 1;

        var recipe = recipes.get(currentRecipe);
        switch (recipe) {
            case CookingRecipe<?> cooking -> renderCookingRecipe(cooking);
            case ShapelessRecipe shapeless -> renderShapelessRecipe(shapeless);
            case ShapedRecipe shaped -> renderShapedRecipe(shaped);
            case TransmuteRecipe transmute -> renderTransmuteRecipe(transmute);
            case SmithingRecipe smithing -> renderSmithingTransformRecipe(smithing);
            case StonecuttingRecipe stonecutting -> renderStonecuttingRecipe(stonecutting);
            case ComplexRecipe complex -> renderComplexRecipe(complex);  // Not really necessary, this is more of a quirk if anything...
            default ->
                    SMPRPG.broadcastToOperatorsCausedBy(this.player, ComponentUtils.create("Unknown recipe handler: " + recipe.getClass().getSimpleName()));
        }
    }

    private void renderComplexRecipe(ComplexRecipe complex) {

        setSlot(CORNER + 10, InterfaceUtil.getNamedItemWithDescription(
                Material.BARRIER,
                ComponentUtils.create("Complex Recipe", NamedTextColor.RED),
                ComponentUtils.EMPTY,
                ComponentUtils.create("This recipe is considered 'complex',"),
                ComponentUtils.create("which means that we can't"),
                ComponentUtils.create("determine the process to craft the"),
                ComponentUtils.create("result. There's not much we can provide"),
                ComponentUtils.create("you in this interface about this recipe!"),
                ComponentUtils.EMPTY,
                ComponentUtils.merge(ComponentUtils.create("Recipe Key: "), ComponentUtils.create(complex.getKey().asString(), NamedTextColor.RED))
        ));

    }

    private void renderStonecuttingRecipe(StonecuttingRecipe stonecutting) {
        var input = getItemFromRecipeChoice(stonecutting.getInputChoice());
        setButton(CORNER + 10, input, event -> handleIngredientClick(input));
        setSlot(CORNER + 18 + 1, InterfaceUtil.getNamedItem(Material.STONECUTTER, ComponentUtils.create("Stonecutter Recipe", NamedTextColor.GOLD)));
    }

    private void renderSmithingTransformRecipe(SmithingRecipe smithing) {

        if (smithing instanceof SmithingTransformRecipe transform) {
            var input = getItemFromRecipeChoice(transform.getTemplate());
            if (input.getType().equals(Material.AIR))
                input = InterfaceUtil.getNamedItem(Material.BARRIER, ComponentUtils.create("No template needed!", NamedTextColor.GREEN));
            ItemStack finalInput = input;
            setButton(CORNER + 9, input, event -> handleIngredientClick(finalInput));
        }

        var base = getItemFromRecipeChoice(smithing.getBase());
        var addition = getItemFromRecipeChoice(smithing.getAddition());

        if (addition.getType().equals(Material.AIR))
            addition = InterfaceUtil.getNamedItem(Material.BARRIER, ComponentUtils.create("No additional item needed!", NamedTextColor.GREEN));

        if (base.getType().equals(Material.AIR))
            base = InterfaceUtil.getNamedItem(Material.BARRIER, ComponentUtils.create("No base item needed!", NamedTextColor.GREEN));

        ItemStack finalBase = base;
        ItemStack finalAddition = addition;
        setButton(CORNER + 10, base, event -> handleIngredientClick(finalBase));
        setButton(CORNER + 11, addition, event -> handleIngredientClick(finalAddition));

        setSlot(CORNER + 19, InterfaceUtil.getNamedItem(Material.SMITHING_TABLE, ComponentUtils.create("Smithing Recipe", NamedTextColor.GOLD)));
    }

    private void renderShapelessRecipe(ShapelessRecipe shapeless) {
        int x = 0;
        int y = 0;

        // Loop through all the choices. These are essentially just the ingredients.
        for (var choice : shapeless.getChoiceList()) {

            var item = getItemFromRecipeChoice(choice);
            this.setButton(y*9+x+CORNER, item, event -> handleIngredientClick(item));

            x += 1;
            // If we are out of bounds, go to the next row.
            if (x >= 3) {
                x = 0;
                y += 1;
            }

        }
        this.setSlot(CORNER + 27 + 1, InterfaceUtil.getNamedItem(Material.CRAFTING_TABLE, ComponentUtils.create("Shapeless Crafting Recipe", NamedTextColor.GOLD)));
    }

    private void renderShapedRecipe(ShapedRecipe shaped) {
        int x = 0;
        int y = 0;
        for (String row : shaped.getShape()) {
            for (char ingredient : row.toCharArray()) {
                var choice = shaped.getChoiceMap().get(ingredient);
                var item = getItemFromRecipeChoice(choice);
                this.setButton(y*9+x+CORNER, item, event -> handleIngredientClick(item));
                x+= 1;
            }
            y += 1;
            x = 0;
        }
        this.setSlot(CORNER + 27 + 1, InterfaceUtil.getNamedItem(Material.CRAFTING_TABLE, ComponentUtils.create("Shaped Crafting Recipe", NamedTextColor.GOLD)));
    }

    /**
     * Renders a transmute recipe on the interface.
     * Transmute recipes are pretty simple, it is just two items that turn an item into another, similar to shapeless.
     * @param transmute The transmute recipe.
     */
    private void renderTransmuteRecipe(TransmuteRecipe transmute) {
        var input = getItemFromRecipeChoice(transmute.getInput());
        var transmuter = getItemFromRecipeChoice(transmute.getMaterial());
        this.setButton(CORNER + 9 + 1, input, event -> handleIngredientClick(input));
        this.setButton(CORNER + 9 + 2, transmuter, event -> handleIngredientClick(transmuter));
        this.setSlot(CORNER + 27 + 1, InterfaceUtil.getNamedItem(Material.CRAFTING_TABLE, ComponentUtils.create("Transmute Crafting Recipe", NamedTextColor.GOLD)));
    }

    private void renderCookingRecipe(CookingRecipe<?> cooking) {

        var top = CORNER + 1;
        var middle = top + 9;
        var bottom = middle + 9;

        ItemStack smelter;

        if (cooking instanceof FurnaceRecipe)
            smelter = InterfaceUtil.getNamedItem(Material.FURNACE, ComponentUtils.create("Furnace Recipe", NamedTextColor.GOLD));
        else if (cooking instanceof BlastingRecipe)
            smelter = InterfaceUtil.getNamedItem(Material.BLAST_FURNACE, ComponentUtils.create("Blasting Recipe", NamedTextColor.GOLD));
        else if (cooking instanceof SmokingRecipe)
            smelter = InterfaceUtil.getNamedItem(Material.SMOKER, ComponentUtils.create("Smoker Recipe", NamedTextColor.GOLD));
        else if (cooking instanceof CampfireRecipe)
            smelter = InterfaceUtil.getNamedItem(Material.CAMPFIRE, ComponentUtils.create("Campfire Recipe", NamedTextColor.GOLD));
        else
            smelter = InterfaceUtil.getNamedItem(Material.BARRIER, ComponentUtils.create("Unknown Smelter: " + cooking.getClass().getSimpleName(), NamedTextColor.RED));

        smelter.lore(ComponentUtils.cleanItalics(List.of(
                ComponentUtils.EMPTY,
                ComponentUtils.merge(ComponentUtils.create("Cook in a "), Component.translatable(smelter.translationKey()).color(NamedTextColor.YELLOW), ComponentUtils.create(" block")),
                ComponentUtils.merge(ComponentUtils.create("for "), ComponentUtils.create(cooking.getCookingTime() / 20 + "s", NamedTextColor.GREEN))
        )));

        var ingredient = getItemFromRecipeChoice(cooking.getInputChoice());
        this.setButton(top, ingredient, event -> handleIngredientClick(ingredient));
        this.setSlot(middle, smelter);
        this.setSlot(bottom, InterfaceUtil.getNamedItem(Material.COAL, ComponentUtils.EMPTY));
    }

    /**
     * Generates an item stack that shows what "requirements" an item has in order to craft. This is subject to change,
     * but for now it simply displays what item "discovers" the recipe upon picking up. A common example is how the
     * recipe for a copper chestplate is discovered when we pick up a copper ingot.
     *
     * @param requirements A collection of item stacks that is "required" to discover this recipe.
     * @return an item stack to be used as a display in the interface
     */
    public ItemStack getRequirements(Collection<ItemStack> requirements) {

        ItemStack paper = createNamedItem(Material.PAPER, ComponentUtils.create("Crafting Requirements", NamedTextColor.RED));
        paper.editMeta(meta -> {
            List<Component> lore = new ArrayList<>();
            lore.add(ComponentUtils.EMPTY);
            lore.add(ComponentUtils.create("The following items must be"));
            lore.add(ComponentUtils.create("discovered to unlock this recipe"));
            lore.add(ComponentUtils.EMPTY);
            for (ItemStack item : requirements)
                lore.add(ComponentUtils.create("- ").append(item.displayName()));
            meta.lore(ComponentUtils.cleanItalics(lore));
            meta.setEnchantmentGlintOverride(true);
        });
        SMPRPG.getService(ItemService.class).setIgnoreMetaUpdate(paper);
        return paper;
    }

    private void changePage(int delta) {
        currentRecipe += delta;

        if (delta < 0)
            this.sounds.playPagePrevious();
        else
            this.sounds.playPageNext();

        render();
    }

    public void render() {

        this.clear();
        this.setBorderFull();

        this.renderRecipe();

        // Misc buttons
        this.setSlot(RESULT, result);

        var blueprint = ItemService.blueprint(result);
        if (blueprint instanceof ICraftable craftable)
            this.setSlot(REQUIREMENTS, getRequirements(craftable.unlockedBy()));

        if (recipes.size() > 1) {
            this.setButton(37, InterfaceUtil.getNamedItem(Material.ARROW, ComponentUtils.create("Previous Page " + (currentRecipe+1) + "/" + recipes.size(), NamedTextColor.GOLD)), event -> changePage(-1));
            this.setButton(39, InterfaceUtil.getNamedItem(Material.ARROW, ComponentUtils.create("Next Page " + (currentRecipe+1) + "/" + recipes.size(), NamedTextColor.GOLD)), event -> changePage(1));
        }

        if (this.player.isOp()) {
            String key = "?";
            if (recipes.get(currentRecipe) instanceof Keyed keyed)
                key = keyed.getKey().asString();
            this.setSlot(0, InterfaceUtil.getNamedItemWithDescription(
                    Material.OAK_SIGN,
                    ComponentUtils.create("Info", NamedTextColor.RED),
                    ComponentUtils.create("Key: " + key, NamedTextColor.YELLOW)
            ));
        }

        // Utility buttons
        this.setButton((ROWS-1)*9+4, BUTTON_BACK, (e) -> this.openParentMenu());
    }
}

File: src/main/java/xyz/devvydont/smprpg/gui/items/MenuRecipeViewer.kt
package xyz.devvydont.smprpg.gui.items

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.TextComponent
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.Bukkit
import org.bukkit.Keyed
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryOpenEvent
import org.bukkit.inventory.*
import org.bukkit.inventory.RecipeChoice.ExactChoice
import org.bukkit.inventory.RecipeChoice.MaterialChoice
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.scheduler.BukkitTask
import xyz.devvydont.smprpg.SMPRPG
import xyz.devvydont.smprpg.SMPRPG.Companion.broadcastToOperatorsCausedBy
import xyz.devvydont.smprpg.SMPRPG.Companion.plugin
import xyz.devvydont.smprpg.gui.InterfaceUtil.getNamedItem
import xyz.devvydont.smprpg.gui.InterfaceUtil.getNamedItemWithDescription
import xyz.devvydont.smprpg.gui.base.MenuBase
import xyz.devvydont.smprpg.gui.base.MenuButtonClickHandler
import xyz.devvydont.smprpg.items.interfaces.ICraftable
import xyz.devvydont.smprpg.services.ItemService
import xyz.devvydont.smprpg.services.ItemService.Companion.blueprint
import xyz.devvydont.smprpg.services.ItemService.Companion.clean
import xyz.devvydont.smprpg.services.ItemService.Companion.generate
import xyz.devvydont.smprpg.services.RecipeService.Companion.getRecipesFor
import xyz.devvydont.smprpg.util.formatting.ComponentUtils
import xyz.devvydont.smprpg.util.time.TickTime
import java.util.List
import java.util.function.Consumer

class MenuRecipeViewer(
    player: Player,
    parentMenu: MenuBase?,
    private val recipes: MutableList<Recipe>,
    result: ItemStack
) : MenuBase(player, ROWS, parentMenu) {
    // The index of the recipe we want to show if there is more than one recipe.
    private var currentRecipe = 0

    private val result: ItemStack =
        clean(result) // This may seem silly, but it's necessary to remove the "craftable" lore.

    // Integer that will allow us to continuously flip through different options a recipe provides.
    private var recipeChoiceIndex = 0

    override fun handleInventoryOpened(event: InventoryOpenEvent) {
        super.handleInventoryOpened(event)
        event.titleOverride(ComponentUtils.merge(ComponentUtils.create("Recipes for: "), result.displayName()))
        this.render()
        Bukkit.getScheduler().runTaskTimer(plugin, Consumer runTaskTimer@{ task: BukkitTask ->

            // If nobody is viewing us, we can stop the task.
            if (this.inventory.viewers.isEmpty()) {
                task.cancel()
                return@runTaskTimer
            }

            render()
            recipeChoiceIndex++
        }, TickTime.HALF_SECOND, TickTime.HALF_SECOND)
    }

    override fun handleInventoryClicked(event: InventoryClickEvent) {
        super.handleInventoryClicked(event)
        event.isCancelled = true
        this.playInvalidAnimation()
    }

    private fun getItemFromRecipeChoice(choice: RecipeChoice?): ItemStack {
        // Yes, this is possible... LMAO
        if (choice == null) return ItemStack.of(Material.AIR)

        var item: ItemStack? = null
        if (choice is ExactChoice)
            item = clean(choice.choices[recipeChoiceIndex % choice.choices.size])

        if (choice is MaterialChoice)
            item = generate(choice.choices[recipeChoiceIndex % choice.choices.size])

        // This will only occur if Spigot/Paper ever add a new child of RecipeChoice in a future update.
        if (item == null) {
            broadcastToOperatorsCausedBy(
                this.player,
                ComponentUtils.create("Unknown recipe choice candidate: " + choice.javaClass.getSimpleName())
            )
            throw IllegalStateException("Unknown choice type: " + choice.javaClass.getSimpleName())
        }

        blueprint(item).updateItemData(item)

        // If this item is deemed craftable by us, inject lore explaining that they can click it to go deeper.
        if (!getRecipesFor(item).isEmpty()) {
            val lore = ArrayList<Component>(
                listOf<TextComponent>(
                    ComponentUtils.EMPTY,
                    ComponentUtils.create("Click to view recipe!", NamedTextColor.YELLOW),
                    ComponentUtils.EMPTY
                )
            )
            if (item.lore() != null) lore.addAll(item.lore()!!)
            item.lore(lore)
        }

        return item
    }

    /**
     * When an ingredient is clicked, if the blueprint of that ingredient is craftable itself as well then open
     * another layer of this menu with its recipe.
     *
     * @param itemStack The ItemStack clicked as a crafting component
     */
    private fun handleIngredientClick(itemStack: ItemStack) {
        // Retrieve the blueprint of this item. If it is craftable, enter another recipe layer

        val recipesFor = getRecipesFor(blueprint(itemStack).generate())
        if (recipesFor.isEmpty() || itemStack.type == Material.AIR) {
            this.playInvalidAnimation()
            return
        }

        this.openSubMenu(MenuRecipeViewer(player, this, recipesFor, itemStack))
    }

    /**
     * Renders the recipe completely
     */
    fun renderRecipe() {
        if (recipes.isEmpty()) {
            broadcastToOperatorsCausedBy(
                this.player,
                ComponentUtils.create(
                    "Achieved impossible recipe viewer state (no recipes to show): ",
                    NamedTextColor.RED
                ).append(result.displayName())
            )
            return
        }

        if (currentRecipe >= recipes.size)
            currentRecipe = 0
        if (currentRecipe < 0)
            currentRecipe = recipes.size - 1
        
        when (val recipe = recipes[currentRecipe]) {
            is CookingRecipe<*> -> renderCookingRecipe(recipe)
            is ShapelessRecipe -> renderShapelessRecipe(recipe)
            is ShapedRecipe -> renderShapedRecipe(recipe)
            is TransmuteRecipe -> renderTransmuteRecipe(recipe)
            is SmithingTransformRecipe -> renderSmithingTransformRecipe(recipe)
            is StonecuttingRecipe -> renderStonecuttingRecipe(recipe)
            is ComplexRecipe -> renderComplexRecipe(recipe)
            else -> broadcastToOperatorsCausedBy(
                this.player,
                ComponentUtils.create("Unknown recipe handler: " + recipe.javaClass.getSimpleName())
            )
        }
    }

    private fun renderComplexRecipe(complex: ComplexRecipe) {
        setSlot(
            CORNER + 10, getNamedItemWithDescription(
                Material.BARRIER,
                ComponentUtils.create("Complex Recipe", NamedTextColor.RED),
                ComponentUtils.EMPTY,
                ComponentUtils.create("This recipe is considered 'complex',"),
                ComponentUtils.create("which means that we can't"),
                ComponentUtils.create("determine the process to craft the"),
                ComponentUtils.create("result. There's not much we can provide"),
                ComponentUtils.create("you in this interface about this recipe!"),
                ComponentUtils.EMPTY,
                ComponentUtils.merge(
                    ComponentUtils.create("Recipe Key: "),
                    ComponentUtils.create(complex.key.asString(), NamedTextColor.RED)
                )
            )
        )
    }

    private fun renderStonecuttingRecipe(stonecutting: StonecuttingRecipe) {
        val input = getItemFromRecipeChoice(stonecutting.inputChoice)
        setButton(
            CORNER + 10,
            input
        ) { event: InventoryClickEvent -> handleIngredientClick(input) }
        setSlot(
            CORNER + 18 + 1,
            getNamedItem(Material.STONECUTTER, ComponentUtils.create("Stonecutter Recipe", NamedTextColor.GOLD))
        )
    }

    private fun renderSmithingTransformRecipe(smithing: SmithingRecipe) {
        if (smithing is SmithingTransformRecipe) {
            var input = getItemFromRecipeChoice(smithing.template)
            if (input.type == Material.AIR) input =
                getNamedItem(Material.BARRIER, ComponentUtils.create("No template needed!", NamedTextColor.GREEN))
            val finalInput = input
            setButton(
                CORNER + 9,
                input
            ) { event: InventoryClickEvent -> handleIngredientClick(finalInput) }
        }

        var base = getItemFromRecipeChoice(smithing.base)
        var addition = getItemFromRecipeChoice(smithing.addition)

        if (addition.type == Material.AIR) addition =
            getNamedItem(Material.BARRIER, ComponentUtils.create("No additional item needed!", NamedTextColor.GREEN))

        if (base.type == Material.AIR) base =
            getNamedItem(Material.BARRIER, ComponentUtils.create("No base item needed!", NamedTextColor.GREEN))

        val finalBase = base
        val finalAddition = addition
        setButton(
            CORNER + 10,
            base
        ) { event: InventoryClickEvent -> handleIngredientClick(finalBase) }
        setButton(
            CORNER + 11,
            addition
        ) { event: InventoryClickEvent -> handleIngredientClick(finalAddition) }

        setSlot(
            CORNER + 19,
            getNamedItem(Material.SMITHING_TABLE, ComponentUtils.create("Smithing Recipe", NamedTextColor.GOLD))
        )
    }

    private fun renderShapelessRecipe(shapeless: ShapelessRecipe) {
        var x = 0
        var y = 0

        // Loop through all the choices. These are essentially just the ingredients.
        for (choice in shapeless.getChoiceList()) {
            val item = getItemFromRecipeChoice(choice)
            this.setButton(
                y * 9 + x + CORNER,
                item
            ) { event: InventoryClickEvent -> handleIngredientClick(item) }

            x += 1
            // If we are out of bounds, go to the next row.
            if (x >= 3) {
                x = 0
                y += 1
            }
        }
        this.setSlot(
            CORNER + 27 + 1,
            getNamedItem(
                Material.CRAFTING_TABLE,
                ComponentUtils.create("Shapeless Crafting Recipe", NamedTextColor.GOLD)
            )
        )
    }

    private fun renderShapedRecipe(shaped: ShapedRecipe) {
        var x = 0
        var y = 0
        for (row in shaped.shape) {
            for (ingredient in row.toCharArray()) {
                val choice = shaped.getChoiceMap()[ingredient]
                val item = getItemFromRecipeChoice(choice)
                this.setButton(
                    y * 9 + x + CORNER,
                    item
                ) { event: InventoryClickEvent -> handleIngredientClick(item) }
                x += 1
            }
            y += 1
            x = 0
        }
        this.setSlot(
            CORNER + 27 + 1,
            getNamedItem(Material.CRAFTING_TABLE, ComponentUtils.create("Shaped Crafting Recipe", NamedTextColor.GOLD))
        )
    }

    /**
     * Renders a transmute recipe on the interface.
     * Transmute recipes are pretty simple, it is just two items that turn an item into another, similar to shapeless.
     * @param transmute The transmute recipe.
     */
    private fun renderTransmuteRecipe(transmute: TransmuteRecipe) {
        val input = getItemFromRecipeChoice(transmute.input)
        val transmuter = getItemFromRecipeChoice(transmute.material)
        this.setButton(
            CORNER + 9 + 1,
            input
        ) { event: InventoryClickEvent -> handleIngredientClick(input) }
        this.setButton(
            CORNER + 9 + 2,
            transmuter
        ) { event: InventoryClickEvent -> handleIngredientClick(transmuter) }
        this.setSlot(
            CORNER + 27 + 1,
            getNamedItem(
                Material.CRAFTING_TABLE,
                ComponentUtils.create("Transmute Crafting Recipe", NamedTextColor.GOLD)
            )
        )
    }

    private fun renderCookingRecipe(cooking: CookingRecipe<*>) {
        val top: Int = CORNER + 1
        val middle = top + 9
        val bottom = middle + 9

        val smelter = when (cooking) {
            is FurnaceRecipe -> getNamedItem(Material.FURNACE, ComponentUtils.create("Furnace Recipe", NamedTextColor.GOLD))
            is BlastingRecipe -> getNamedItem(Material.BLAST_FURNACE, ComponentUtils.create("Blasting Recipe", NamedTextColor.GOLD))
            is SmokingRecipe -> getNamedItem(Material.SMOKER, ComponentUtils.create("Smoker Recipe", NamedTextColor.GOLD))
            is CampfireRecipe -> getNamedItem(Material.CAMPFIRE, ComponentUtils.create("Campfire Recipe", NamedTextColor.GOLD))
            else -> getNamedItem(
                Material.BARRIER,
                ComponentUtils.create("Unknown Smelter: " + cooking.javaClass.getSimpleName(), NamedTextColor.RED)
            )
        }

        smelter.lore(
            ComponentUtils.cleanItalics(
                listOf<Component>(
                    ComponentUtils.EMPTY,
                    ComponentUtils.merge(
                        ComponentUtils.create("Cook in a "),
                        Component.translatable(smelter.translationKey()).color(NamedTextColor.YELLOW),
                        ComponentUtils.create(" block")
                    ),
                    ComponentUtils.merge(
                        ComponentUtils.create("for "),
                        ComponentUtils.create((cooking.cookingTime / 20).toString() + "s", NamedTextColor.GREEN)
                    )
                )
            )
        )

        val ingredient = getItemFromRecipeChoice(cooking.inputChoice)
        this.setButton(
            top,
            ingredient
        ) { event: InventoryClickEvent -> handleIngredientClick(ingredient) }
        this.setSlot(middle, smelter)
        this.setSlot(bottom, getNamedItem(Material.COAL, ComponentUtils.EMPTY))
    }

    /**
     * Generates an item stack that shows what "requirements" an item has in order to craft. This is subject to change,
     * but for now it simply displays what item "discovers" the recipe upon picking up. A common example is how the
     * recipe for a copper chestplate is discovered when we pick up a copper ingot.
     *
     * @param requirements A collection of item stacks that is "required" to discover this recipe.
     * @return an item stack to be used as a display in the interface
     */
    fun getRequirements(requirements: MutableCollection<ItemStack>): ItemStack {
        val paper = createNamedItem(Material.PAPER, ComponentUtils.create("Crafting Requirements", NamedTextColor.RED))
        paper.editMeta(Consumer { meta: ItemMeta ->
            val lore: MutableList<Component?> = ArrayList()
            lore.add(ComponentUtils.EMPTY)
            lore.add(ComponentUtils.create("The following items must be"))
            lore.add(ComponentUtils.create("discovered to unlock this recipe"))
            lore.add(ComponentUtils.EMPTY)
            for (item in requirements) lore.add(ComponentUtils.create("- ").append(item.displayName()))
            meta.lore(ComponentUtils.cleanItalics(lore))
            meta.setEnchantmentGlintOverride(true)
        })
        SMPRPG.getService(ItemService::class.java).setIgnoreMetaUpdate(paper)
        return paper
    }

    private fun changePage(delta: Int) {
        currentRecipe += delta

        if (delta < 0) this.sounds.playPagePrevious()
        else this.sounds.playPageNext()

        render()
    }

    fun render() {
        this.clear()
        this.setBorderFull()

        this.renderRecipe()

        // Misc buttons
        this.setSlot(RESULT, result)

        val blueprint = blueprint(result)
        if (blueprint is ICraftable) this.setSlot(REQUIREMENTS, getRequirements(blueprint.unlockedBy()))

        if (recipes.size > 1) {
            this.setButton(
                37,
                getNamedItem(
                    Material.ARROW,
                    ComponentUtils.create(
                        "Previous Page " + (currentRecipe + 1) + "/" + recipes.size,
                        NamedTextColor.GOLD
                    )
                )
            ) { event: InventoryClickEvent -> changePage(-1) }
            this.setButton(
                39,
                getNamedItem(
                    Material.ARROW,
                    ComponentUtils.create(
                        "Next Page " + (currentRecipe + 1) + "/" + recipes.size,
                        NamedTextColor.GOLD
                    )
                )
            ) { event: InventoryClickEvent -> changePage(1) }
        }

        if (this.player.isOp) {
            var key = "?"
            val recipe = recipes[currentRecipe]
            if (recipe is Keyed)
                key = (recipe as Keyed).key.asString()
            this.setSlot(
                0, getNamedItemWithDescription(
                    Material.OAK_SIGN,
                    ComponentUtils.create("Info", NamedTextColor.RED),
                    ComponentUtils.create("Key: $key", NamedTextColor.YELLOW)
                )
            )
        }

        // Utility buttons
        this.setButton(
            (ROWS - 1) * 9 + 4,
            BUTTON_BACK
        ) { e: InventoryClickEvent -> this.openParentMenu() }
    }

    companion object {
        const val ROWS: Int = 6

        // Hard set positions
        const val CORNER: Int = 10
        const val RESULT: Int = 23
        const val REQUIREMENTS: Int = 25
    }
}
